internal/types/secret.go

```diff
@@ -23,200 +22,308 @@ import (
+		// Nothing to redact
 	case *schema.GoModulesConnection:
-		return []jsonStringField{}, nil
+		for i := range c.Urls {
+			if c.Urls[i] == "" {
+				continue
+			}
+
+			redacted, err := redactedURL(c.Urls[i])
+			if err != nil {
+				return "", err
+			}
+
+			edits = append(edits, edit{jsonx.MakePath("urls", i), redacted})
+		}
 	case *schema.JVMPackagesConnection:
-		return []jsonStringField{{[]string{"maven", "credentials"}, &cfg.Maven.Credentials}}, nil
+		if c.Maven != nil && c.Maven.Credentials != "" {
+			edits = append(edits, edit{jsonx.MakePath("maven", "credentials"), redactedString(c.Maven.Credentials)})
+		}
 	case *schema.PagureConnection:
-		if cfg.Token != "" {
-			return []jsonStringField{{[]string{"token"}, &cfg.Token}}, nil
+		if c.Token != "" {
+			edits = append(edits, edit{jsonx.MakePath("token"), redactedString(c.Token)})
 		}
-		return []jsonStringField{}, nil
 	case *schema.NpmPackagesConnection:
-		return []jsonStringField{{[]string{"credentials"}, &cfg.Credentials}}, nil
+		if c.Credentials != "" {
+			edits = append(edits, edit{jsonx.MakePath("credentials"), redactedString(c.Credentials)})
+		}
 	case *schema.OtherExternalServiceConnection:
-		return []jsonStringField{{[]string{"url"}, &cfg.Url}}, nil
+		if c.Url != "" {
+			redacted, err := redactedURL(c.Url)
+			if err != nil {
+				return "", err
+			}
+			edits = append(edits, edit{jsonx.MakePath("url"), redacted})
+		}
 	default:
 		// return an error; it's safer to fail than to incorrectly return unsafe data.
-		return nil, errors.Errorf("Unrecognized ExternalServiceConfig for redaction: kind %+v not implemented", reflect.TypeOf(cfg))
+		return "", errors.Errorf("Unrecognized ExternalServiceConfig for redaction: kind %+v not implemented", reflect.TypeOf(cfg))
+	}
+
+	redacted := e.Config
+	for _, e := range edits {
+		if redacted, err = e.apply(redacted); err != nil {
+			return "", err
+		}
 	}
+
+	return redacted, nil
 }
 
-// UnredactConfig will replace redacted fields with their undredacted form from the 'old' ExternalService.
+// UnredactConfig will replace redacted fields with their unredacted form from the 'old' ExternalService.
 // You should call this when accepting updated config from a user that may have been
 // previously redacted, and pass in the unredacted form directly from the DB as the 'old' parameter
 func (e *ExternalService) UnredactConfig(old *ExternalService) error {
 	if e == nil || old == nil || e.Config == "" || old.Config == "" {
 		return nil
 	}
+
 	if old.Kind != e.Kind {
 		return errors.Errorf(
 			"UnRedactExternalServiceConfig: unmatched external service kinds, old: %q, e: %q",
 			old.Kind,
 			e.Kind,
 		)
 	}
-	var (
-		unredacted string
-		err        error
-	)
-	cfg, err := e.Configuration()
+
+	newCfg, err := e.Configuration()
 	if err != nil {
 		return err
 	}
-	unredacted, err = unredactFields(old.Config, e.Config, cfg)
+
+	oldCfg, err := old.Configuration()
 	if err != nil {
 		return err
 	}
+
+	var edits []edit
```

Comment:
Suggestion (non-blocking): A small helper function would help compress the repetitive code a bit.

```suggestion
	var edits []edit
	redactIfNonempty := func (new, old string, path string...) {
	    if newValue != "" {
	      edits = append(edits, edit{jsonx.MakePath(path...), unredactedString(new, old))
	    }
	}
```

Would help compress the code vertically quite a bit.

Reply:
Addressed in the latest commit.